pkg/controller/master/virtualmachine/vmi_network_controller.go

```diff
@@ -12,107 +13,179 @@ import (
-	if err != nil {
-		return err
+func getVMInterfaceMacs(vm *kubevirtv1.VirtualMachine) map[string]string {
+	vmInterfaceMacs := make(map[string]string, len(vm.Spec.Template.Spec.Domain.Devices.Interfaces))
+	for _, vmIface := range vm.Spec.Template.Spec.Domain.Devices.Interfaces {
+		if vmIface.MacAddress != "" {
+			vmInterfaceMacs[vmIface.Name] = vmIface.MacAddress
+		}
 	}
+	return vmInterfaceMacs
+}
 
-	vmCopy := vm.DeepCopy()
-	var length = len(vmi.Status.Interfaces)
-	var vmiInterfaces = make(map[string]string, length)
+func getVmiInterfaceMacs(vmi *kubevirtv1.VirtualMachineInstance) map[string]string {
+	var vmiInterfaces = make(map[string]string, len(vmi.Status.Interfaces))
 	for _, iface := range vmi.Status.Interfaces {
-		if iface.MAC != "" && iface.Name != "" {
+		if iface.Name != "" && iface.MAC != "" {
 			vmiInterfaces[iface.Name] = iface.MAC
 		}
 	}
+	return vmiInterfaces
+}
 
-	for i, vmIface := range vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces {
-		macAddress, ok := vmiInterfaces[vmIface.Name]
-		// only set the network mac address when it has no existing value
-		if ok && vmIface.MacAddress == "" {
-			logrus.Debugf("set VM %s management network %s macAddress to %s", vm.Name, vmIface.Name, macAddress)
-			vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces[i].MacAddress = macAddress
-		}
+func (h *VMNetworkController) syncControllerRevision(vmi *kubevirtv1.VirtualMachineInstance, retry *bool) (*kubevirtv1.VirtualMachineInstance, error) {
+	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return vmi, err
 	}
 
-	// note: this function is related to vm/vmi, should always run on vmi change
-	if err := h.regenerateControllerRevision(vmi, vm); err != nil {
-		return err
+	if vm == nil || vm.DeletionTimestamp != nil {
+		return vmi, nil
 	}
 
-	if equality.Semantic.DeepEqual(vmCopy.Spec.Template.Spec.Domain.Devices, vm.Spec.Template.Spec.Domain.Devices) {
-		return nil
+	// already synced
+	if vm.Generation == vm.Status.ObservedGeneration {
+		return vmi, nil
 	}
-	if _, err := h.vmClient.Update(vmCopy); err != nil {
-		return err
+
+	if err := h.regenerateControllerRevision(vmi, vm, retry); err != nil {
+		return vmi, err
 	}
 
-	return nil
+	return vmi, nil
 }
 
-func (h *VMNetworkController) regenerateControllerRevision(vmi *kubevirtv1.VirtualMachineInstance, vm *kubevirtv1.VirtualMachine) error {
+func (h *VMNetworkController) regenerateControllerRevision(vmi *kubevirtv1.VirtualMachineInstance, vm *kubevirtv1.VirtualMachine, retry *bool) error {
 	crObj, err := h.crCache.Get(vmi.Namespace, vmi.Status.VirtualMachineRevisionName)
 	if err != nil {
-		return fmt.Errorf("error fetch controller revision object for vmi %s-%s: %v", vmi.Name, vmi.Namespace, err)
+		// controllerRevision has a different name than vm name, essential to log it's name in error
+		return fmt.Errorf("failed to fetch controllerRevision %s/%s: %w", vmi.Namespace, vmi.Status.VirtualMachineRevisionName, err)
```

Comment:
IIUC, even with this PR, there is still a chance that we can encounter the `not found, requeuing` error if the cache isn't sync'ed. Is it worth doing retrying the `GET` here, in case the error is transient? The error also seems to imply some requeuing is happening.

Reply:
in rancher warngler & k8s controller mode,  most controller does not loop wait a cache to be synced in the normal run time, that will block current go thread and nobody knows exactly how long/how many times(e.g. try 3 times get) to wait;  instead, the controller reconciller mechanism ensure this, after return error, the controller will be retriggered. a returned & logged error like `not found` is normal in the pod